storage/src/merkle_tree_db_adapter.cpp

```diff
@@ -328,25 +342,30 @@ Status DBAdapter::getBlockById(BlockId blockId, Sliver &block) const {
   for (const auto &[key, keyData] : blockNode.keys) {
     if (!keyData.deleted) {
       Sliver value;
-      if (const auto status = db_->get(DBKeyManipulator::genDataDbKey(key, blockId), value); !status.isOK()) {
+      if (const auto status = db_->get(DBKeyManipulator::genDataDbKey(key, blockNode.stateRootVersion.value()), value);
+          !status.isOK()) {
         // If the key is not found, treat as corrupted storage and abort.
         Assert(!status.isNotFound());
         return status;
       }
-      keyValues[key] = value;
+      const auto dbLeafVal = detail::deserialize<detail::DatabaseLeafValue>(value);
+      Assert(dbLeafVal.blockId == blockId);
+      keyValues[key] = dbLeafVal.leafNode.value;
     }
   }
 
   block = block::create(blockId, keyValues, blockNode.parentDigest.data(), blockNode.stateHash);
   return Status::OK();
 }
 
-concordUtils::SetOfKeyValuePairs DBAdapter::lastReachableBlockkDbUpdates(const SetOfKeyValuePairs &updates,
-                                                                         BlockId blockId) {
-  const auto updateBatch = smTree_.update(updates);
-
-  // Key updates.
-  auto dbUpdates = batchToDbUpdates(updateBatch);
+SetOfKeyValuePairs DBAdapter::lastReachableBlockkDbUpdates(const SetOfKeyValuePairs &updates, BlockId blockId) {
+  auto dbUpdates = SetOfKeyValuePairs{};
+  // Create a block with the same state root as the previous one if there are no updates.
+  if (!updates.empty()) {
+    // Key updates.
+    const auto updateBatch = smTree_.update(updates);
+    dbUpdates = batchToDbUpdates(updateBatch, blockId);
+  }
 
   // Block key.
   dbUpdates[DBKeyManipulator::genBlockDbKey(blockId)] = createBlockNode(updates, blockId, smTree_.get_version());
```

Comment:
Looking at this I was wondering why we didn't also pass `smTree.get_root_hash(`). It turns out that we don't need it in `createBlockNode` since `smTree_` is a member variable of the class. Therefore we also don't need to pass `smTree_.get_version()` here either. This can be made in a separate change, but I think it makes the code clearer.

Reply:
Great catch! Doing that.